src/main/java/implementations2/MyBigInteger.java

```
@@ -0,0 +1,415 @@
+
+/**
+ * Created by VickyD on 10/26/15.
+ * //todo
+ * convert some static methods to instance methods
+ * subtract()
+ */
+public class MyBigInteger implements Comparable<MyBigInteger> {
+
+
+    private final String inputString;
+//    private  char sign;
+//    private  String unsignedString;
+//    public  int[] value;
+
+    public MyBigInteger(String str) {
+        validNumberCheck(str);
+        inputString = str;
+    }
+
+    public String toString() {
+        String result = getUnsignedString(this.inputString);
+        result = trimLeadingZeroes(result);
+        if(result.length() == 0) return "0";
+
+        if(this.isSignNegative()) return "-" + result;
+        return result;
+
+    }
+
+    public boolean equals(MyBigInteger big2) {
+        if (this == big2) {
+            return true;
+        } else {
+            //check for sign and value
+            char sign1 = this.getSign();
+            char sign2 = big2.getSign();
+            String unsignedString1 = getUnsignedString(this.inputString);
+            String unsignedString2 = getUnsignedString(big2.inputString);
+            if ((sign1 == sign2) && unsignedString1.equals(unsignedString2)) {
+                return true;
+            }
+        }
+        return false;
+
+    }
+
+    public int hashCode() {
+        Character sign = getSign();
+        String unsignedString = getUnsignedString(inputString);
+        return sign.hashCode()*31 + unsignedString.hashCode();
+
+    }
+
+    @Override
+    public int compareTo(MyBigInteger big2) {
+        if (this.equals(big2)) return 0;
+        if (bothArePositive(this, big2)) return compareMagnitude(big2);
+        if (bothAreNegative(this, big2)) return compareMagnitude(big2) * (-1);
+        if (this.isSignPositive()) return +1;
+        if (this.isSignNegative()) return -1;
+        return 0;  //this is not required.. check : Last statement could be above stmt without if i.e. return -1; but this is more readable code
```

Then throw ``IllegalStateException`` here, even without any comment every developer would understand, that the intention that it never riches this statement (i.e. everything should happen above). Also it could save in case of "bad" refactoring (e.g. something was changed in the lines above or in the methods which is called from the lines above, and the intention is broken, then it'll start fail with ``IllegalStateException`` instead of silently mark two ``MyBigInteger`` as equal (``return 0;``).